llvm/lib/CAS/CachingOnDiskFileSystem.cpp

```diff
@@ -289,34 +364,86 @@ CachingOnDiskFileSystemImpl::makeFile(DirectoryEntry &Parent,
 }
 
 Expected<FileSystemCache::DirectoryEntry *>
-CachingOnDiskFileSystemImpl::lookupOnDiskFrom(DirectoryEntry &Parent,
-                                              StringRef Name) {
+CachingOnDiskFileSystemImpl::lookupOnDiskFrom(
+    DirectoryEntry &Parent, const std::string &TreePath,
+    Optional<sys::fs::file_status> KnownStatus) {
   assert(Parent.isDirectory() && "Expected a directory");
 
-  // Open the path on disk.
-  StringRef ParentPath = Parent.getTreePath();
-  std::string Path =
-      (ParentPath.endswith("/") ? ParentPath + Name : ParentPath + "/" + Name)
-          .str();
-
   // lstat is extremely slow...
   sys::fs::file_status Status;
-  if (std::error_code EC = sys::fs::status(Path, Status, /*follow=*/false))
-    return errorCodeToError(EC);
+  if (KnownStatus) {
+    Status = std::move(*KnownStatus);
+  } else {
+    if (auto EC = sys::fs::status(TreePath, Status, /*follow=*/false))
+      return errorCodeToError(EC);
+  }
 
   if (Status.type() == sys::fs::file_type::directory_file)
-    return makeDirectory(Parent, Path);
+    return makeDirectory(Parent, TreePath);
 
   if (Status.type() == sys::fs::file_type::symlink_file)
-    return makeSymlink(Parent, Path);
+    return makeSymlink(Parent, TreePath);
 
   int FD;
   if (std::error_code EC =
-          sys::fs::openFile(Path, FD, sys::fs::CD_OpenExisting,
+          sys::fs::openFile(TreePath, FD, sys::fs::CD_OpenExisting,
                             sys::fs::FA_Read, sys::fs::OF_None))
     return errorCodeToError(EC);
   auto CloseOnExit = make_scope_exit([&FD]() { ::close(FD); });
-  return makeFile(Parent, Path, FD, Status);
+  return makeFile(Parent, TreePath, FD, Status);
+}
+
+Expected<FileSystemCache::DirectoryEntry *>
+CachingOnDiskFileSystemImpl::lookupOnDiskFrom(DirectoryEntry &Parent,
+                                              StringRef Name,
+                                              DirectoryEntry *KnownRealPath) {
+  assert(Parent.isDirectory() && "Expected a directory");
+
+  StringRef ParentPath = Parent.getTreePath();
+  std::string Path =
+      (ParentPath.endswith("/") ? ParentPath + Name : ParentPath + "/" + Name)
+          .str();
+
+  // lstat is extremely slow...
+  sys::fs::file_status Status;
+  if (std::error_code EC = sys::fs::status(Path, Status, /*follow=*/false))
+    return errorCodeToError(EC);
+
+  DirectoryEntry *Next =
+      KnownRealPath ? KnownRealPath->nextEntryAfterPrefix(Parent) : nullptr;
+  if (!Next) {
+    // We do not know the realpath, or it does not contain Parent, which
+    // may indicate we are in the middle of looking up a path that contains a
+    // symlink to an absolute path in the current or a subsequent component, or
+    // that we hit F_GETPATH non-determinism for a hard link on Darwin. We need
+    // to get the realpath of Parent + Name.
+    auto RP = preloadRealPath(Parent, Name);
```

Comment:
I don't think this call to `preloadRealPath()` will get cached because it's from the `CachingOnDiskFileSystemImpl` instead of the `CachingOnDiskFileSystemDI` (`DiscoveryInstanceImpl`). Should this function take one of the latter as an argument, or maybe be moved to a helper function on it? (or, did I misread the logic? is it okay like this?)

Reply:
Oh, I think I did misread it. Can you add "[... to get the realpath of Parent + Name], which is different from the realpath of the file that's ultimately being opened by the calling DiscoveryInstance" (or something to that effect).

I still wonder if this ought to live in DiscoveryInstance; might be more clear that `FS.preload` is being called on purpose, and it's only ever called from there so it might make it more obvious what the context is. Not sure though.